README
/**************************STELLARINO USER GUIDE*************************/
                Copyright (C) 2012-2014 Sultan Qasim Khan

Stellarino is free software; you can redistribute it and/or modify it
under the terms of the GNU Lesser General Public License as published by
the Free Software Foundation; either version 3 of the License, or
(at your option) any later version.

Stellarino is distributed in the hope that it will be useful, but WITHOUT
ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
FITNESS FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public
License for more details.

You should have received a copy of the GNU Lesser General Public License
along with Stellarino. If not, see <http://www.gnu.org/licenses/>.


/***************************ABOUT STELLARINO***************************/
Stellarino is a simple Wiring-like but incompatible interface library
implemented in C for popular TI Stellaris and Tiva C series microntrollers,
particularly those used in Launchpad boards. It is meant to facilitate the
rapid creation of basic microcontroller firmware without the need to
parse through thousands of pages in datasheets and other documentation.
It uses syntax similar to Wiring to ease the introduction of newcomers
familiar with the Arduino and similar boards to the Stellaris platform.

Stellarino implements functions for configuration, delays, using timers,
GPIO, analog input, PWM, controlling hobby-style servomotors, UART, and
more. It was created using the TI StellarisWare Peripheral Driver Library.
It thus requires StellarisWare and a suitable IDE, such as TI Code Composer
Studio in order to function. Since Stellarino implements all basic
functionality using the StellarisWare PDL with simple, easy to understand
code, it also makes a good introductory example for use of the PDL.


/***********************SUPPORTED MICROCONTROLLERS**********************/
- TI Stellaris LM4F120H5QR (Stellaris Launchpad)
- TI Tiva C Series TM4C123GH6PM (Tiva C Series Launchpad)
- Support for the TM4C1294NCPDT used in the Tiva C Series Connected
  Launchpad is coming soon


/************************HOW TO BUILD STELLARINO************************/
On Linux and Mac:
See gcc_instructions.txt

On Windows:
In order to build Stellarino, you must first build the TivaWare Peripheral
Peripheral Driver Library for the Cortex M4F, contained within the CCS
project driverlib, and have that project inside your CCS workspace.

If you have not already done that, first do as follows:
1. In CCS, go to Project->Import Existing CCS Eclipse Project
2. Beside "Select search-directory" click on "Browse"
3. Browse to the root of your StellarisWare installation directory, select
   the folder called "driverlib", and hit OK. Two projects should be
   discovered in the box below.
4. Tick the project "driverlib" and make sure "Copy projects into workspace"
   is selected. Hit finish.
5. Open the driverlib project and hit build. You should be building the
   debug configuration.

Once driverlib has been built within one's CCS workspace, one may
proceed to import and build Stellarino. To build a release version of
Stellarino, you will need to build a release version of driverlib.


/***************************PIN MULTIPLEXING***************************/
Pin     Possible Use(s) in Stellarino
PA0     GPIO, UART0 RX
PA1     GPIO, UART0 TX
PA2     GPIO, SSI0 SCK
PA3     GPIO, SSI0 SS
PA4     GPIO, SSI0 MISO
PA5     GPIO, SSI0 MOSI
PA6     GPIO
PA7     GPIO
PB0     GPIO, PWM, UART1 RX
PB1     GPIO, PWM, UART1 TX
PB2     GPIO, PWM
PB3     GPIO, PWM
PB4     GPIO, PWM, ADC, SSI2 SCK
PB5     GPIO, PWM, ADC, SSI2 SS
PB6     GPIO, PWM, SSI2 MISO
PB7     GPIO, PWM, SSI2 MOSI
PC0     JTAG (Reserved)
PC1     JTAG (Reserved)
PC2     JTAG (Reserved)
PC3     JTAG (Reserved)
PC4     GPIO, PWM, UART4 RX
PC5     GPIO, PWM, UART4 TX
PC6     GPIO, PWM, UART3 RX
PC7     GPIO, PWM, UART3 TX
PD0     GPIO, PWM, ADC, SSI3 SCK
PD1     GPIO, PWM, ADC, SSI3 SS
PD2     GPIO, PWM, ADC, SSI3 MISO
PD3     GPIO, PWM, ADC, SSI3 MOSI
PD4     GPIO, UART6 RX
PD5     GPIO, UART6 TX
PD6     GPIO, UART2 RX
PD7     GPIO, UART2 TX
PE0     GPIO, ADC
PE1     GPIO, ADC
PE2     GPIO, ADC
PE3     GPIO, ADC
PE4     GPIO, ADC, UART5 RX
PE5     GPIO, ADC, UART5 TX
PE6     GPIO, UART7 RX
PE7     GPIO, UART7 TX
PF0     GPIO, PWM, SSI1 MISO    // Also defined as SW2
PF1     GPIO, PWM, SSI1 MOSI    // Also defined as RED_LED
PF2     GPIO, PWM, SSI1 SCK     // Also defined as BLUE_LED
PF3     GPIO, PWM, SSI1 SS      // Also defined as GREEN_LED
PF4     GPIO, PWM               // Also defined as SW1
PF5     GPIO
PF6     GPIO
PF7     GPIO


/**************************GENERAL FUNCTIONS**************************/
void init(void);
    This function must be called at the beginning of main() in any program
    to configure the microcontroller to function with Stellarino.


/**********************INPUT AND OUTPUT FUNCTIONS*********************/
NOTE: 	FOR ALL IO FUNCTIONS, PIN SHOULD BE SPECIFIED IN THE FORM PX#.
        For example, Port A Pin 0 would be PA0. Pin names in this form
        have been predefined as macros in the library.

void pinMode(unsigned char pin, unsigned char mode);
    Configures the given pin for the given mode.
    mode should be one of the following (self-explanatory) macros:
        INPUT
        OUTPUT
        OUTPUT_OD       // Open drain GPIO
        INPUT_PULLUP
        INPUT_PULLDOWN
        INPUT_ANALOG
        OUTPUT_PWM      // 500 Hz PWM, customizable via stellarino.h
        OUTPUT_PWM_OD   // Open drain PWM
        OUTPUT_SERVO    // Standard servo pulses every 20 ms

int digitalRead(unsigned char pin);
    Reads a pin configured as a digital input and returns 1 or 0
    depending on whether the pin is HIGH or LOW.

int analogRead(unsigned char pin);
    Uses the ADC to read the analog voltage on a pin configured for
    analog input. Returns a value from 0 to 4095. Reading takes 13 to
    23 microseconds.

void digitalWrite(unsigned char pin, short val);
    Writes a value to a pin configured for digital output.
    Please note that writing HIGH to an input pin will not give it
    pullup resistance. For pullup and pulldown, use the appropriate
    setting in pinMode.
    val can either be the macros HIGH or LOW (or the numbers 1 or 0)

void analogWrite(unsigned char pin, short val);
    Sets the PWM duty cycle of a pin configured for PWM output.
    val is the duty cycle, from 0 to 255. 0 is always off, 255 is
    always on.

void servoWrite(unsigned char pin, short val);
    Sets the pulse length in microseconds for a pin configured for servo
    pulse output.
    val is the pulse length in microseconds. val must be between 600 and
        2400 microseconds.

void pwmWrite(unsigned char pin, float frequency, float duty);
    Sets a pin to output PWM at the given frequency with the given duty
    cycle. The pin must first be set up for PWM output through pinMode().
    After using this function on a pin, you may not use analogWrite on the
    same pin until pinMode is called to restore the pin configuration.
    frequency is the PWM output frequency in Hz.
    duty is the duty cycle. It must be a value between 0.0 and 1.0.

unsigned long pulseIn(unsigned char pin, short val, unsigned long timeout);
    Waits for a pulse to begin on a GPIO input pin until timeout, and
    returns the length of the pulse in microseconds if a pulse is detected.
    This function supports pulses with lengths between 1 us and 7 minutes.
    val is the polarity of the pulse, HIGH or LOW (1 or 0)
    timeout is the time to wait for a pulse to begin, in microseconds


/***************************TIMING FUNCTIONS**************************/
void delay(unsigned long nTime);
    Pause for nTime milliseconds. Supports pauses up to 5 days in length.

void delayMicroseconds(unsigned long nTime);
    Pause for nTime microseconds.

void resetMillis(void);
    Reset a timer that will measure elapsed time in milliseconds.

unsigned long millis(void);
    Returns the elapsed time in milliseconds since the millisecond timer was
    last reset. This timer can count for up to almost 25 days.

void resetMicros(void);
unsigned long micros(void);
    Similar to above, but time is measured in microseconds instead of
    milliseconds. This timer can only count for 71 minutes.


/****************************UART FUNCTIONS***************************/
Functions without the UART prefix use UART0, which is connected to the ICDI
on the Launchpad. UART-prefixed functions can use any UART interface, from
UART0 all the way to UART7. Just specify the appropriate UART number as an
unsigned char. For example, to write a character to UART3, use the command
UARTputc(3, mychar);

Data on the ICDI UART is transmitted at 115200 baud by default, with 8 data
bits and 1 stop bit. The baud rate can be changed through a call to the
enableUART function.

void enableUART(uint8_t UART, unsigned long baudRate);
    Enables the given UART, and configures it to run at the given baud rate.
    Any arbitrary baud rate may be used, however for reliable communication,
    baud rates of 1200000 or below are recommended.
    
int16_t UARTgetBufferLevel(uint8_t UART);
    Returns the number of bytes present in the UART receive buffer.

bool UARToverflow(uint8_t UART);
    Returns true if the UART receive buffer has overflowed.

void UARTflushReceiveBuffer(uint8_t UART);
    Flushes all data in the UART receive buffer and resets the overflow flag.

void puts(const char * str);
void UARTputs(uint8_t UART, const char * str);
    Outputs the string str over UART. Please note that unlike the standard
    C implementation of puts, this DOES NOT automatically append a newline.

char * gets(char * str, int num);
char * UARTgets(uint8_t UART, char * str, int num);
    Reads a string over UART until a newline is detected or the requested
    number of characters have been read.
    str is a pointer to a character array (C-style string) to which the
        string is read.
    num is the maximum number of characters to read (including the string
        termination character '\0')

void putc(char c);
void UARTputc(uint8_t UART, char c);
    Outputs the character c over UART.

void putln(void);
void UARTputln(uint8_t UART);
    Outputs a CRLF newline over UART.

char getc(void);
char UARTgetc(uint8_t UART);
    Returns a character read over UART. Waits for a character if one is not
    yet available. The character is erased from the buffer once read.

int peek(void);
int UARTpeek(uint8_t UART);
    Returns the next character from the buffer without clearing it.
    Successive calls to peek() and peekBlocking, and the next call to getc()
    will return the same character. Returns -255 if the buffer is empty.

char peekBlocking(void);
char UARTpeekBlocking(uint8_t UART);
    Returns the next character from the buffer without clearing it.
    Successive calls to peek() and peekBlocking, and the next call to getc()
    will return the same character. Waits for a character if the buffer is
    empty.

void puti(long i);
void UARTputi(uint8_t UART, long i);
    Outputs the signed long integer i as a decimal.

long geti(void);
long UARTgeti(uint8_t UART);
    Gets a decimal signed long integer over UART and returns it.

void putu(unsigned long u, uint8_t digits);
void UARTputu(uint8_t UART, unsigned long u, uint8_t digits);
    Outputs the unsigned long integer u as a decimal. Adds leading zeroes
    or truncates as required to match given number of digits.
    digits is the number of digits desired.

unsigned long getu(uint8_t digits);
unsigned long UARTgetu(uint8_t UART, uint8_t digits);
    Gets a decimal unsigned long integer over UART and returns it. Reads up
    to a non numeric character or till the digit limit.
    digits is the maximum number of digits to read.

void puth(unsigned long h, uint8_t digits);
void UARTputh(uint8_t UART, unsigned long h, uint8_t digits);
    Outputs the unsigned long integer h as a hexadecimal. Adds leading
    zeroes or truncates as required to match given number of digits.
    digits is the number of digits desired.

unsigned long geth(uint8_t digits);
unsigned long UARTgeth(uint8_t UART, uint8_t digits);
    Gets a hexadecimal unsigned long integer over UART and returns it.
    Reads up to a non numeric character or till the digit limit.
    digits is the maximum number of digits to read.

void putf(float f, uint8_t decimal);
void UARTputf(uint8_t UART, float f, uint8_t decimal);
    Outputs the floating point number f to the given number of decimal
    places over UART.
    decimal is the number of decimal places to output.

float getf();
float UARTgetf(uint8_t UART);
    Gets a floating point number over UART and returns it. This function
    does not accept numbers in scientific notation.


/***************************SPI/SSI FUNCTIONS*************************/
void enableSPI(unsigned short SPINum, unsigned short wordLength, unsigned long dataRate);
    Enables the selected SPI, and configures it for a specified
    word length in bits at a data rate specified in Hz. The pins associated
    with the selected SPI will be configured for SPI.
    SPINum is the number of the SPI (0 to 3)
    wordLenth is the length of each word in bits
    dataRate is the transmission rate in Hz

void SPIWrite(unsigned short SPINum, unsigned long data);
    Writes the given data to the selected SPI.
    SPINum is the number of the SPI
    data is the word to be transmitted

unsigned long SPIRead(unsigned short SPINum);
    Reads a word from the selected SPI. Will wait for data if buffer empty.
    SPINum is the number of the SPI

changelog.txt
/**************************STELLARINO CHANGELOG**************************/
               Copyright (C) 2012-2013 Sultan Qasim Khan

Stellarino is free software; you can redistribute it and/or modify it
under the terms of the GNU Lesser General Public License as published by
the Free Software Foundation; either version 3 of the License, or
(at your option) any later version.

Stellarino is distributed in the hope that it will be useful, but WITHOUT
ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
FITNESS FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public
License for more details.

You should have received a copy of the GNU Lesser General Public License
along with Stellarino. If not, see <http://www.gnu.org/licenses/>.
/************************************************************************/

Version History:
0.1     Dec 14, 2012    Initial version. Basic init() function, delay() and
                        delayMicroseconds(), basic GPIO functions, i.e.
                        pinMode(), digitalWrite(), digitalRead().

0.2     Dec 15, 2012    Power-saving interrupt based delay() function that
                        puts processor to sleep during a delay implemented.
                        The improved delay function allows delays up to almost
                        5 days length. Crude pulseIn() function implemented.

0.3     Dec 17, 2012    PWM output implemented with the analogWrite() function.
                        Servo signal PWM mode added, and servoWrite() function
                        created for interfacing. Much improved implementation
                        of pulseIn() created using a timer instead of empirical
                        timing.

0.4     Dec 18, 2012    Implemented analogRead() and analog pin configuration.
                        Added timer functions using WTIMER4 to track elapsed time.
                        Broke off pin multiplexing and other global constant arrays
                        into stellarino_pins.h for better organization. Delay
                        functions grouped with timer functions in stellarino_timer.h
                        Switched PDL commands to ROM versions to reduce binary size.

0.5     Dec 19, 2012    Implemented basic FOSS UART library to eliminate dependence
                        on incompatibly licensed UARTstdio library. Renamed timer
                        functions to millis() and micros() to better match Wiring.
                        Modified timer functions to count one unit of time per timer
                        tick instead of one per ten ticks. This increased the range
                        of the timers ten-fold. Timers are now started automatically
                        upon init() to better match Wiring, but the ability to reset
                        the timers has been maintained. Fixed delayMicroseconds().
                        Added example using the setup() and loop() structure of Wiring
                        sketches.

0.6     Dec 20, 2012    Increased clock speed to 80 MHz and adjusted timer setup
                        appropriately. Added UART functions for unsigned decimal and
                        hexadecimal input and output.

0.7     Dec 26, 2012    Added SPI (master) support. Increased stack size to 512 bytes
                        to permit more complex programs and the safe use of printf and
                        similarly complex functions. Add MCP4921 DAC demo using SPI.
                        Set up release compile configuration. Added code to prevent
                        bricking through accidentally reconfiguring the JTAG pins.

0.8     Jan 1, 2013     Fixed 0 duty cycle PWM. Added instructions to README for
                        building driverlib-cm4f. Added post-build step to create a
                        flashable .bin file. Added vector table to permit code to run
                        automatically after a reset.
                        
0.9     Apr 7, 2013     Fixed millis() timer. Further increased stack size to 1024
                        bytes as there is no shortage of RAM on this MCU. Added new
                        UART functions for using UART ports other than UART0.

1.0     Jun 14, 2013    Fixed multiple PWM outputs from one timer. Created new two
                        speed LED colour fade example as the new default sample
                        program. Reduced PWM frequency to 500 Hz. The PWM frequency
                        is customizable via a macro in stellarino.h. Added servo
                        support to all PWM-capable pins. Added enableUART function
                        to simplify the configuration of UARTs.

1.1     Jul 10, 2013    Fixed geti/getu/geth functions for UART1 and up. Added pin
                        lock code to unlock PD7 for use as UART2. Created new UART
                        functions for floating point input and output. Added map()
                        macro for linearly mapping a number from one range to another.
                        
1.2     Oct 20, 2013    Switched to LGPL licence. Added UART RX buffering. Fixed bug
                        in putf that could cause stack overflow and infinite loop.
                        Removed math.h dependency. Improved project structure and
                        code formatting. Clarified the return value for UARTpeek.

1.3     Nov 28, 2013    Added support for Tiva C Series Launchpad. Added support for
                        development on Linux and Mac using GCC for ARM. Fixed
                        critical UART receive buffering bug (thanks to astro75).
                        Added new pwmWrite write function for arbitrary PWM output.
                        Added experimental support for using stdio.h functions such
                        as printf and and sscanf. Tested only with GCC. Some functions
                        are presently broken, such as scanf. Switched to TivaWare

stellarino.c
/*  stellarino.c
    Copyright (C) 2012-2014 Sultan Qasim Khan

    This is part of Stellarino.

    Stellarino is free software; you can redistribute it and/or modify it
    under the terms of the GNU Lesser General Public License as published by
    the Free Software Foundation; either version 3 of the License, or
    (at your option) any later version.

    Stellarino is distributed in the hope that it will be useful, but WITHOUT
    ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
    FITNESS FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public
    License for more details.

    You should have received a copy of the GNU Lesser General Public License
    along with Stellarino. If not, see <http://www.gnu.org/licenses/>.
*/

#include <stellarino.h>

static const long pwmPeriod = 80000000 / PWMFREQ;

void init(void)
{
    // Set system clock to 80 MHz
    ROM_SysCtlClockSet(SYSCTL_SYSDIV_2_5|SYSCTL_USE_PLL|SYSCTL_XTAL_16MHZ|SYSCTL_OSC_MAIN);

    // Enable FPU and allow floating point operations in interrupts
    ROM_FPUEnable();
    ROM_FPULazyStackingEnable();

    // Enable the UART connected to the ICDI on the Launchpad
    // UART will run at 115200 baud, 8 data bits, 1 stop bit
    enableUART(0, 115200);

    // Configure WTIMER4 for Timer functions
    ROM_SysCtlPeripheralEnable(SYSCTL_PERIPH_WTIMER4);
    ROM_TimerConfigure(WTIMER4_BASE, (TIMER_CFG_SPLIT_PAIR | TIMER_CFG_A_ONE_SHOT | TIMER_CFG_B_ONE_SHOT));
    ROM_TimerPrescaleSet(WTIMER4_BASE, TIMER_A, 39999);         // 0.5 ms per cycle
    ROM_TimerPrescaleSet(WTIMER4_BASE, TIMER_B, 79);            // 1 us per cycle
    ROM_SysCtlPeripheralSleepEnable(SYSCTL_PERIPH_WTIMER4);
    ROM_TimerEnable(WTIMER4_BASE, TIMER_A);
    ROM_TimerEnable(WTIMER4_BASE, TIMER_B);

    // Configure WTIMER5 for delay() function
    ROM_SysCtlPeripheralEnable(SYSCTL_PERIPH_WTIMER5);
    ROM_TimerConfigure(WTIMER5_BASE, (TIMER_CFG_SPLIT_PAIR | TIMER_CFG_A_ONE_SHOT | TIMER_CFG_B_ONE_SHOT));
    ROM_TimerPrescaleSet(WTIMER5_BASE, TIMER_A, 7999);          // 0.1 ms per cycle
    ROM_TimerPrescaleSet(WTIMER5_BASE, TIMER_B, 7);             // 0.1 us per cycle
    TimerIntRegister(WTIMER5_BASE, TIMER_A, delayInterrupt);    // Attach interrupt to function
    ROM_IntMasterEnable();	// Enable interrupts
    ROM_SysCtlPeripheralSleepEnable(SYSCTL_PERIPH_WTIMER5);
}

void pinMode(PinNumber pin, GPIOPinMode mode)
{
    // Anti-brick JTAG Protection
    if (pin >= PC0 && pin <= PC3) return;

    ROM_SysCtlPeripheralEnable(SysCtlGPIOs[pin/8]);
    ROM_SysCtlPeripheralSleepEnable(SysCtlGPIOs[pin/8]);

    switch (mode)
    {
    case INPUT:
        ROM_GPIOPinTypeGPIOInput(GPIO[pin/8], bit8[pin%8]);
        ROM_GPIOPadConfigSet(GPIO[pin/8], bit8[pin%8], GPIO_STRENGTH_8MA, GPIO_PIN_TYPE_STD);
        break;

    case OUTPUT:
        ROM_GPIOPinTypeGPIOOutput(GPIO[pin/8], bit8[pin%8]);
        ROM_GPIOPadConfigSet(GPIO[pin/8], bit8[pin%8], GPIO_STRENGTH_8MA, GPIO_PIN_TYPE_STD);
        break;
        
    case OUTPUT_OD:
        ROM_GPIOPinTypeGPIOOutputOD(GPIO[pin/8], bit8[pin%8]);
        ROM_GPIOPadConfigSet(GPIO[pin/8], bit8[pin%8], GPIO_STRENGTH_8MA, GPIO_PIN_TYPE_OD);
        break;

    case INPUT_PULLUP:
        ROM_GPIOPinTypeGPIOInput(GPIO[pin/8], bit8[pin%8]);
        ROM_GPIOPadConfigSet(GPIO[pin/8], bit8[pin%8], GPIO_STRENGTH_8MA, GPIO_PIN_TYPE_STD_WPU);
        break;

    case INPUT_PULLDOWN:
        ROM_GPIOPinTypeGPIOInput(GPIO[pin/8], bit8[pin%8]);
        ROM_GPIOPadConfigSet(GPIO[pin/8], bit8[pin%8], GPIO_STRENGTH_8MA, GPIO_PIN_TYPE_STD_WPD);
        break;

    case INPUT_ANALOG:
        ROM_SysCtlPeripheralEnable(SYSCTL_PERIPH_ADC0);
        ROM_GPIOPinTypeADC(GPIO[pin/8], bit8[pin%8]);
        break;

    case OUTPUT_PWM:
    case OUTPUT_PWM_OD:
        if (pinMux[pin][0] == 12) break;
        ROM_SysCtlPeripheralEnable(SysCtlTimers[pinMux[pin][0]]);
        ROM_SysCtlPeripheralSleepEnable(SysCtlTimers[pinMux[pin][0]]);

        ROM_GPIOPinTypeTimer(GPIO[pin/8], bit8[pin%8]);
        ROM_GPIOPinConfigure(pinMux[pin][2]);

        if (mode == OUTPUT_PWM_OD)
            ROM_GPIOPadConfigSet(GPIO[pin/8], bit8[pin%8], GPIO_STRENGTH_8MA, GPIO_PIN_TYPE_OD);
        else
            ROM_GPIOPadConfigSet(GPIO[pin/8], bit8[pin%8], GPIO_STRENGTH_8MA, GPIO_PIN_TYPE_STD);

        ROM_TimerConfigure(TIMER[pinMux[pin][0]], (TIMER_CFG_SPLIT_PAIR | TIMER_CFG_A_PWM | TIMER_CFG_B_PWM));

        // The narrow timers may need to use a prescaler to get the desired PWM period
        if (pinMux[pin][0] < 6)
            ROM_TimerPrescaleSet(TIMER[pinMux[pin][0]], pinMux[pin][1], (pwmPeriod & 0xFFFF0000) >> 16);
        else ROM_TimerPrescaleSet(TIMER[pinMux[pin][0]], pinMux[pin][1], 0);

        // Timer will load this value on timeout
        ROM_TimerLoadSet(TIMER[pinMux[pin][0]], pinMux[pin][1], pwmPeriod);

        // Initial duty cycle of 0
        ROM_TimerMatchSet(TIMER[pinMux[pin][0]], pinMux[pin][1], pwmPeriod - 1);
        if (pinMux[pin][0] < 6)
            ROM_TimerPrescaleMatchSet(TIMER[pinMux[pin][0]], pinMux[pin][1],
                    (pwmPeriod & 0xFFFF0000) >> 16);
        else ROM_TimerPrescaleMatchSet(TIMER[pinMux[pin][0]], pinMux[pin][1], 0);

        ROM_TimerControlLevel(TIMER[pinMux[pin][0]], pinMux[pin][1], 0);
        ROM_TimerEnable(TIMER[pinMux[pin][0]], TIMER_BOTH);
        break;

    case OUTPUT_SERVO:
        if (pinMux[pin][0] == 12) break;
        ROM_SysCtlPeripheralEnable(SysCtlTimers[pinMux[pin][0]]);
        ROM_SysCtlPeripheralSleepEnable(SysCtlTimers[pinMux[pin][0]]);
        ROM_GPIOPadConfigSet(GPIO[pin/8], bit8[pin%8], GPIO_STRENGTH_8MA, GPIO_PIN_TYPE_STD);
        ROM_GPIOPinTypeTimer(GPIO[pin/8], bit8[pin%8]);
        ROM_GPIOPinConfigure(pinMux[pin][2]);
        ROM_TimerConfigure(TIMER[pinMux[pin][0]], (TIMER_CFG_SPLIT_PAIR | TIMER_CFG_A_PWM | TIMER_CFG_B_PWM));

        // The narrow timers may need to use a prescaler to get the desired PWM period
        if (pinMux[pin][0] < 6)
            ROM_TimerPrescaleSet(TIMER[pinMux[pin][0]], pinMux[pin][1], (1600000 & 0xFFFF0000) >> 16);
        else ROM_TimerPrescaleSet(TIMER[pinMux[pin][0]], pinMux[pin][1], 0);

        // Timer will load this value on timeout
        ROM_TimerLoadSet(TIMER[pinMux[pin][0]], pinMux[pin][1], 1600000);

        // Initial duty cycle of 1500 us
        ROM_TimerMatchSet(TIMER[pinMux[pin][0]], pinMux[pin][1], 1480000);
        if (pinMux[pin][0] < 6)
            ROM_TimerPrescaleMatchSet(TIMER[pinMux[pin][0]], pinMux[pin][1],
                    (1480000 & 0xFFFF0000) >> 16);
        else ROM_TimerPrescaleMatchSet(TIMER[pinMux[pin][0]], pinMux[pin][1], 0);

        ROM_TimerControlLevel(TIMER[pinMux[pin][0]], pinMux[pin][1], 0);
        ROM_TimerEnable(TIMER[pinMux[pin][0]], TIMER_BOTH);
        break;
    }
}

int digitalRead(PinNumber pin)
{
    if (ROM_GPIOPinRead(GPIO[pin/8], bit8[pin%8])) return 1;
    else return 0;
}

int analogRead(PinNumber pin)
{
    if (pinMux[pin][3] == 12) return 0;	// Non-ADC pin

    ROM_ADCSequenceConfigure(ADC0_BASE, 0, ADC_TRIGGER_PROCESSOR, 0);
    ROM_ADCSequenceStepConfigure(ADC0_BASE, 0, 0, ADC_CTL_IE | ADC_CTL_END | pinMux[pin][3]);
    ROM_ADCSequenceEnable(ADC0_BASE, 0);

    ROM_ADCProcessorTrigger(ADC0_BASE, 0);
    while(!ROM_ADCIntStatus(ADC0_BASE, 0, false));
    uint32_t value;
    ROM_ADCSequenceDataGet(ADC0_BASE, 0, &value);

    return value;	// From 0 to 4095
}

void digitalWrite(PinNumber pin, short val)
{
    if (val) ROM_GPIOPinWrite(GPIO[pin/8], bit8[pin%8], bit8[pin%8]);
    else ROM_GPIOPinWrite(GPIO[pin/8], bit8[pin%8], 0);
}

void analogWrite(PinNumber pin, short val)
{
    long period;	// Period the output is low

    if (val > 255) period = 0;
    else if (val <= 0) period = pwmPeriod - 1;
    else period = (pwmPeriod * (255-val)) >> 8;

    if (pinMux[pin][0] < 6)	// Narrow timers may need a prescaler
        ROM_TimerPrescaleMatchSet(TIMER[pinMux[pin][0]], pinMux[pin][1], (period & 0xFFFF0000) >> 16);
    ROM_TimerMatchSet(TIMER[pinMux[pin][0]], pinMux[pin][1], period);
}

void servoWrite(PinNumber pin, short val)
{
    long period;

    if (val < 600) val = 600;
    else if (val > 2400) val = 2400;
    period = 1600000 - (80 * val);

    if (pinMux[pin][0] < 6)	// Narrow timers need a prescaler
        ROM_TimerPrescaleMatchSet(TIMER[pinMux[pin][0]], pinMux[pin][1], (period & 0xFFFF0000) >> 16);
    ROM_TimerMatchSet(TIMER[pinMux[pin][0]], pinMux[pin][1], period);
}

void pwmWrite(PinNumber pin, float frequency, float duty)
{
    // First set up the pin for the given frequency
    int customPwmPeriod = 80000000 / frequency;

    // The narrow timers may need to use a prescaler to get the desired PWM period
    if (pinMux[pin][0] < 6)
        ROM_TimerPrescaleSet(TIMER[pinMux[pin][0]], pinMux[pin][1], (customPwmPeriod & 0xFFFF0000) >> 16);
    else ROM_TimerPrescaleSet(TIMER[pinMux[pin][0]], pinMux[pin][1], 0);

    // Timer will load this value on timeout
    ROM_TimerLoadSet(TIMER[pinMux[pin][0]], pinMux[pin][1], customPwmPeriod);

    // Initial duty cycle of 0
    ROM_TimerMatchSet(TIMER[pinMux[pin][0]], pinMux[pin][1], customPwmPeriod - 1);
    if (pinMux[pin][0] < 6)
        ROM_TimerPrescaleMatchSet(TIMER[pinMux[pin][0]], pinMux[pin][1],
                (customPwmPeriod & 0xFFFF0000) >> 16);
    else ROM_TimerPrescaleMatchSet(TIMER[pinMux[pin][0]], pinMux[pin][1], 0);

    // Now set the period the output is low, based on the duty cycle
    long period;
    if (duty >= 1) period = 0;
    else if (duty <= 0) period = customPwmPeriod - 1;
    else period = customPwmPeriod * (1.0f - duty);

    if (pinMux[pin][0] < 6)	// Narrow timers may need a prescaler
        ROM_TimerPrescaleMatchSet(TIMER[pinMux[pin][0]], pinMux[pin][1], (period & 0xFFFF0000) >> 16);
    ROM_TimerMatchSet(TIMER[pinMux[pin][0]], pinMux[pin][1], period);
}

unsigned long pulseIn(PinNumber pin, short val, unsigned long timeout)
{
    // Max supported pulse length is 7 minutes
    ROM_TimerEnable(WTIMER5_BASE, TIMER_B);

    ROM_TimerLoadSet(WTIMER5_BASE, TIMER_B, timeout * 10);
    ROM_SysCtlDelay(5);	// Give time for timer to reload and count down once
    while (digitalRead(pin) != val)
    {
        // Timeout if reload has occurred
        if (ROM_TimerValueGet(WTIMER5_BASE, TIMER_B) == timeout * 10) return 0;
    }

    ROM_TimerLoadSet(WTIMER5_BASE, TIMER_B, 4294967295);
    ROM_SysCtlDelay(5);
    while (digitalRead(pin) == val);
    unsigned long pulseLength = (4294967295 - ROM_TimerValueGet(WTIMER5_BASE, TIMER_B)) / 10;

    ROM_TimerDisable(WTIMER5_BASE, TIMER_B);
    return pulseLength;
}
